utils/neural_magic/sparse_train_manager.py

```diff
@@ -191,6 +194,49 @@ def turn_off_scaler(self, scaler: torch.cuda.amp.GradScaler):
         """
         scaler._enabled = False
 
+    def rescale_gradient_accumulation(
+        self, batch_size: int, accumulate: int, image_size: int
+    ) -> Tuple[int, int]:
+        """
+        Used when autobatch and QAT are both enabled. Training with QAT adds additional
+        overhead which can cause OOM errors if autobatch is enabled. This function
+        rescales batch size and gradient accumulation to fit into memory with QAT while
+        maintaining the original effective batch size
+        """
+        # Temporary copy of the model with QAT applied
+        quant_model_copy = deepcopy(de_parallel(self.model))
+        self.train_manager.apply_structure(quant_model_copy, float("inf"))
+
+        # Calculate maximum batch size that will fit in memory
+        new_batch_size = check_train_batch_size(quant_model_copy, image_size, False)
+
+        # Calculate batch size closest to maximum that can be accumulated to maintain
+        # the original effective batch size. Note that if the original batch size is odd
+        # then the effective batch size will be incremented by 1
+        batch_size = batch_size if batch_size % 2 == 0 else batch_size + 1
+        batch_size_ratio = math.floor(batch_size / new_batch_size)
+        closest_divisor = next(
+            divisor
+            for divisor in range(batch_size_ratio, 1, -1)
+            if batch_size % divisor == 0
+        )
+        new_batch_size = batch_size // closest_divisor
+        new_accumulate = (batch_size * accumulate) // new_batch_size
```

Comment:
This isn't always ideal. e.g. if the original batch size is 21 and the largest we could fit into memory is 7, we wouldn't choose 7 because we'd be operating with 22. I can update to cover those cases, if we still choose to go this route